src/back/hlsl/writer.rs

```diff
@@ -30,16 +50,84 @@ struct EpStructMember {
     pub binding: Option<crate::Binding>,
 }
 
+#[derive(Clone, Copy, Debug, Hash, Eq, Ord, PartialEq, PartialOrd)]
+struct WrappedImageQuery {
+    dim: crate::ImageDimension,
+    arrayed: bool,
+    class: crate::ImageClass,
+    query: ImageQuery,
+}
+
+// HLSL backend requires its own `ImageQuery` enum.
+// It is used inside `WrappedImageQuery` and should be unique per ImageQuery function.
+// IR version can't be unique per function, because it's store mipmap level as an expression.
+//
+// For example:
+// ```wgsl
+// let dim_cube_array_lod = textureDimensions(image_cube_array, 1);
+// let dim_cube_array_lod2 = textureDimensions(image_cube_array, 1);
+// ```
+//
+// ```ir
+// ImageQuery {
+//  image: [1],
+//  query: Size {
+//      level: Some(
+//          [1],
+//      ),
+//  },
+// },
+// ImageQuery {
+//  image: [1],
+//  query: Size {
+//      level: Some(
+//          [2],
+//      ),
+//  },
+// },
+// ```
+//
+// HLSL should generate only 1 function for this case.
+//
+#[derive(Clone, Copy, Debug, Hash, Eq, Ord, PartialEq, PartialOrd)]
+enum ImageQuery {
+    Size,
+    SizeLevel,
+    NumLevels,
+    NumLayers,
+    NumSamples,
+}
+
+impl From<crate::ImageQuery> for ImageQuery {
+    fn from(q: crate::ImageQuery) -> Self {
+        use crate::ImageQuery as Iq;
+        match q {
+            Iq::Size { level } => {
+                if level.is_some() {
+                    ImageQuery::SizeLevel
+                } else {
+                    ImageQuery::Size
+                }
+            }
+            Iq::NumLevels => ImageQuery::NumLevels,
+            Iq::NumLayers => ImageQuery::NumLayers,
+            Iq::NumSamples => ImageQuery::NumSamples,
+        }
+    }
+}
+
 pub struct Writer<'a, W> {
     out: W,
     names: crate::FastHashMap<NameKey, String>,
     namer: proc::Namer,
     /// HLSL backend options
+    #[allow(dead_code)]
     options: &'a Options,
```

Comment:
nit: in cases like this I find renaming this to `_options` to be better. Otherwise, you may start using the options later and forget that there is a decoration put on it.